url04f7fcf4e1_peteratt_HDFS_Replication_Simulator_tgz-pJ8-hdfs_replicationsimulator_MainJ8/src/hdfs/replicationsimulator/Simulator.java

```diff
@@ -0,0 +1,275 @@
+
+		// Create the Namenode
+		initializeNamenode();
+		// Create the datanode distribution and fill the Datanode map
+		initializeDatanodes();
+		// Create the blocks distribution, attribute them to datanodes and fill
+		// the blocksMap
+		initializeBlocks();
+
+	}
+
+	private static void startFailure() {
+		
+		Thread killer = new Thread(new NodeKiller());
+		killer.start();
+		
+	}
+
+	private static void initializeNamenode() {
+		namenode = new Namenode();
+		System.out.print("Namenode Created.\n");
+	}
+
+	private static void initializeDatanodes() {
+
+		// Create all the datanodes
+		allDatanodes = new AllDatanode();
+
+		for (int i = 0; i < numberofDatanodes; i++) {
+			allDatanodes.addNode(new Datanode(i, dataNodeCapacity));
+
+			DatanodeInfo datanodeInfo = new DatanodeInfo(i, dataNodeCapacity);
+			namenode.addNode(datanodeInfo);
+		}
+
+		System.out.print(numberofDatanodes + " Datanodes Created.\n");
+	}
+
+	private static void initializeBlocks() {
+		int currentDN = 0;
+		for (int i = 0; i < numberofBlocks; i++) {
+			// Create the Block
+			Block block = new BlockInfo(i);
+
+			// Add the block to the namenode index
+			namenode.addBlock((BlockInfo)block);
+
+			
+			/* THIS VERSION USE RANDOM AND SO IS NOT GOOD.
+			// Adds the block randomly to a node
+			 * Random r = new Random();
+			for (int j = 0; j < numberofReplicas; j++) {
+				int idDatanode = r.nextInt(numberofDatanodes - 1) ;
+				//int index = idDatanode - 1;
+				Datanode dn = allDatanodes.getNode(idDatanode);//TODO
+				dn.addBlock(block);
+				namenode.initAddBlock(idDatanode, (BlockInfo)block);
+			}
+			*/
+			//adds the blocks to a node (chained mode)
+			
+			for (int j = 0; j < numberofReplicas; j++) {
+				int idDatanode = currentDN;
+				//int index = idDatanode - 1;
+				Datanode dn = allDatanodes.getNode(idDatanode);
+				dn.addBlock(block);
+				namenode.initAddBlock(idDatanode, (BlockInfo)block);
+				
+				currentDN = (currentDN==numberofDatanodes-1)? 0: currentDN+1;
+			}
+		}
+		System.out.print(numberofBlocks + " Blocks distributed\n");
+	}
+
+	public static void start() {
+		allDatanodes.start();
+		namenode.start();
+		startFailure();
+	}
+
+	public static int getBandwidth() {
+		return bandwidth;
+	}
+
+	public static int getHeartbeat() {
+		return heartbeat;
+	}
+
+	public static int getTimeout() {
+		return timeout;
+	}
+
+	public static int getBlockSize() {
+		return blockSize;
+	}
+	
+	public static int getNumberofReplicas() {
+		return numberofReplicas;
+	}
+
+	public static Queue<Event> getToNamenode() {
+		return toNamenode;
+	}
+
+	public static Queue<Event> getToDatanodes() {
+		return toDatanodes;
+	}
+
+	public static Namenode getNamenode() {
+		return namenode;
+	}
+
+	public static void addTrace(SimTrace st) {
+		traceList.add(st);
+		System.out.println(st.toString());
+	}
+
+	public static void printResults() {
+
+		File f = new File("results.log");
+		try {
+			FileWriter fw = new FileWriter(f);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *fw*. Execution paths that do not contain closure statements were detected. Ensure *fw* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).